core/src/test/scala/unit/kafka/coordinator/transaction/ProducerIdManagerTest.scala

```diff
@@ -38,19 +38,57 @@ import org.mockito.Mockito.{mock, when}
 import java.util.concurrent.{CountDownLatch, Executors, TimeUnit}
 import java.util.concurrent.atomic.AtomicBoolean
 import scala.collection.mutable
-import scala.util.{Failure, Success}
+import scala.util.{Failure, Success, Try}
 
 class ProducerIdManagerTest {
 
   var brokerToController: NodeToControllerChannelManager = mock(classOf[NodeToControllerChannelManager])
   val zkClient: KafkaZkClient = mock(classOf[KafkaZkClient])
 
+  case class ErrorCount(error: Errors, var repeat: Int)
+
+  object ErrorCount {
+    val INDEFINITE: Int = -1
+
+    def indefinitely(error: Errors): ErrorCount = {
+      ErrorCount(error, INDEFINITE)
+    }
+  }
+
+  class ErrorQueue(initialErrorCounts: ErrorCount*) {
+    private val queue: mutable.Queue[ErrorCount] = mutable.Queue.empty ++ initialErrorCounts
+
+    def takeError(): Errors = queue.synchronized {
+      while (queue.head.repeat == 0) {
+        queue.dequeue()
+      }
+      if (queue.head.repeat > 0) {
+        queue.head.repeat -= 1
+      }
+      queue.head.error
+    }
+
+    def peekError(): Errors = queue.synchronized {
+      queue.head.error
+    }
+
+    def clearProcessedError(): Unit = {
+      TestUtils.waitUntilTrue(() =>
+        queue.synchronized {
+          queue.head.repeat == 0
+        }, "error wasn't processed")
+      queue.synchronized {
+        queue.dequeue()
```

Comment:
shouldn't we clear the whole queue here?

Reply:
no, but doesn't matter any more. I was able to get rid of the maybeRequestNextBlock override completely and the whole "mocking" became much more easier.